### ISSUE TITLE
`consistent-return` Treat `return void ...` the same as `return;`

### ISSUE THREAD

**Issue by scriptdaemon** [eslint/eslint#6026]
**What version of ESLint are you using?**

2.8.0

**What parser (default, Babel-ESLint, etc.) are you using?**

Default

**Please show your full configuration:**

```
root: true

env:
  es6: true
  node: true

extends:
  "eslint:recommended"

rules:
  consistent-return: "error"
```

**What did you do? Please include the actual source code causing the issue.**

Wrote a function with a callback that returns early when an error is found:

```js
log(level, message, meta, callback) {
  let recipientId = _.get(meta, 'recipientId');
  if (!recipientId) {
    return void callback(new Error('No recipient ID'));
  }

  // Other code that doesn't return anything ...
}
```

Returning the callback is nice because it saves a line, but I want to guarantee the function returns nothing (thus the use of `void`).

**What did you expect to happen?**

No error.

**What actually happened? Please include the actual, raw output from ESLint.**

` error  Expected to return a value at the end of this method  consistent-return`

**Comment by platinumazure** [comment]
Apologies for taking so long to triage this. I have marked this as an enhancement request for now, but if someone else on the team thinks it should be treated as a bug, that person can re-label this issue.

@eslint/eslint-team Should this be regarded as a potential breaking change (and thus the requested functionality should go behind an option)? Or do we think this is safe to implement as a bugfix? (One could argue that we should know that a `void` operation must inherently return void, and so the rule might be misbehaving in that sense.)

**Comment by nzakas** [comment]
We don't calculate values, so even though `void` always returns `undefined`, all the rule knows is that there is a value. So, not a bug, it would be an enhancement. We'd need a champion for this change. I'm 👍 .

**Comment by ilyavolodin** [comment]
Sounds reasonable to me, but I think it should go behind an option. 👍

**Comment by scriptdaemon** [comment]
Any ideas as to what the option should be named?

`ignoreVoid`?

**Comment by pedrottimark** [comment]
@scriptdaemon Could you search the approximate frequency of ignoreX versus allowX options?

It seems like we are starting to use ignore in introduction or Rule Details to mean that something is inherently out of scope for a rule, and allow for options. How does it seem to you?

**Comment by platinumazure** [comment]
@pedrottimark @scriptdaemon In this case, we aren't allowing anything new (`return void x` has always been *allowed*, just marked as a value rather than as void), nor is it being ignored.

I know there is no precedent for this (and surely we can iterate and come up with something better), but I would almost go for something like `implicitVoidOnly` (which is true by default, but can be set to false to allow explicit voids to also be treated as empty returns). Or perhaps `voidIsEmptyReturn` (false by default).

The point I'm getting at is, because we're not doing a simple allow or ignore, I think we need something more explicit.

Thoughts?

**Comment by scriptdaemon** [comment]
@platinumazure I agree. I do like `implicitVoidOnly` myself.

**Comment by pedrottimark** [comment]
Suggest write a draft of item for bullet list under hypothetical This rule has an object option:

* `"X" true` does what [related to either existing or your proposed improvement of rule introduction]

**Comment by pedrottimark** [comment]
avoidExtraLine is tempting, but I doubt it 😂

**Comment by scriptdaemon** [comment]
- `"implicitVoidOnly": false` enforces `void` expressions to be considered equivalent to lone `return` statements

Something like that?

**Comment by nzakas** [comment]
We generally make rule options begin with verbs to make it clear what's going on. `implicitVoidOnly` doesn't tell me what it's doing. It may not be "allow" or "ignore", but it needs a verb to clarify what it's doing.

And I think we need to some cleaner wording. "implicit void" has no real definition (in fact, "implicit undefined" would be more accurate since `void` is an operator that returns `undefined`).

In order to clarify, I'd suggest we expand the scope of this suggestion a bit and consider the following to be equivalent:

```js
return void foo();
return undefined;
return;
```

Effectively, all three return `undefined`, just with different syntax. So I'd suggest an option name of `treatUndefinedAsEmpty` or something similar.

**Comment by scriptdaemon** [comment]
@nzakas I agree with including `return undefined;`, and that name sounds reasonable enough to me.

**Comment by platinumazure** [comment]
Since `undefined` is just a variable, are we also expecting to treat any
declared-but-uninitialized variable the same way? Or should the rule at
least be smart enough to detect `undefined` having been overwritten in an
ancestor scope? (Sorry, I'm just trying to account for corner cases.)
On May 4, 2016 6:50 PM, "Kenneth Williams" <notifications@github.com> wrote:

@nzakas <https://github.com/nzakas> I agree with including return undefined;,
and that name sounds reasonable enough to me.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/eslint/eslint/issues/6026#issuecomment-217041278>

**Comment by scriptdaemon** [comment]
@platinumazure I don't know if it makes sense to me to include checks that returned variables are defined or not, as `return undefined;` and `return void ... ;` are likely intentional, whereas returning an undefined variable might not be intentional (and even if it is in a given case, it might not be in all cases). Your second question makes sense to me, but I wonder if it's outside the scope of the option / rule to make that kind of check.

**Comment by platinumazure** [comment]
I'm not strongly advocating for those ideas. And to be honest, I wouldn't
have suggested them if we were working with a *sane* language that made
`undefined` a keyword. :smile:

**Comment by nzakas** [comment]
@platinumazure I think that's going a bit overboard. There will always be edge cases we can't detect, and that's fine. Let's just start with easy stuff for now.

**Comment by pedrottimark** [comment]
The only part about https://github.com/eslint/eslint/issues/6026#issuecomment-216995837 which leaves me unsure is whether it leaves a way (if there needs to be a way) for a function to be correct according to this rule if it returns `undefined` in some cases and, for example, a number value in others. For lack of a better term, a “polymorphic return value” function. How often does that happen in the wild? I’ve done it before 😳

The rule considers `return undefined` as an **explicit* return value which you could use in that use case (or abuse case ;)

Although `return undefined` has an equivalent **result** to the following:
* code path which reaches end of function
* `return`
* `return void foo()'

I am wondering if the option we are discussing should leave `return undefined` available to express the **intention** of an **explicit return value** category and allow `return void foo()` to become a synonym for the other **implicit return value** alternatives. Is that making any sense?

**Comment by scriptdaemon** [comment]
@pedrottimark If I understand that correctly, you're saying `return void ...` should be treated equal to `return;` (like I suggest in the first post), while `return undefined;` should be equal to `return value;` (in order to express the intent that the function should return *something*)?

**Comment by pedrottimark** [comment]
Yes, because of the sentence under Rule Details:

> This rule is aimed at ensuring all return statements either specify a value or don’t specify a value.

The rule description is even clearer:

> consistent-return: require `return` statements to either always or never specify values

So I think we are back where we started: clearly name and describe an option so `return void foo()` is equivalent to the other ways not to specify a value. My brain fails at this moment. If there are any other idioms, better to consider them now. But it seems like `return undefined` does specify a value.

To bring things into focus, and maybe before we can decide on an option name, we really do need to rewrite the introduction in those terms, maybe with a bulleted list of the ways that a function does not specify a value, similar to https://github.com/eslint/eslint/issues/6026#issuecomment-217194715

@scriptdaemon Could you please write a first draft of an introduction (the content between main heading and Rule Details) for the consistent-return rule?

**Comment by pedrottimark** [comment]
Oh, and I see that the rule docs lacks an example of unspecified! So could you write a minimal realistic code snippet to illustrate what you need for the examples of correct code: a function which does not specify a value, with a structure something like

```
function (args) {
    if (something} {
        // fill in some action here
        return;
    }
}
```

And also its counterpart with `void` for additional correct code?

**Comment by scriptdaemon** [comment]
@pedrottimark I'm having a hard time seeing it that way. To me, a value is *something* (e.g. bool, string, etc.), while `undefined` (and `void`) is explicitly *nothing*. I think this suggestion fits within the existing rule description given that context.

Since this suggestion would go behind an option, I don't think it interferes with a user's intent, either. If they want to specify that a function not have an empty  `return` statement and use `return undefined'`, they can simply not use this option.

Examples for correct code:

```js
/*eslint consistent-return: "error"*/

function doSomething(condition) {
    if (condition) {
        return true;
    } else {
        return false;
    }
}

function Foo() {
    if (!(this instanceof Foo)) {
        return new Foo();
    }

    this.a = 0;
}

function (arg) {
    if (!arg) {
        return;
    }
    // No return
}

function (arg) {
    if (!arg) {
        return undefined;
    }
    // No return
}

function (arg, callback) {
    if (!arg) {
        return void callback(new Error('Oops!'));
    }
    // No return
}
```

**Comment by pedrottimark** [comment]
Yes indeed, finally it clicks: all of those are ways for a function to have consistent return of undefined.

@scriptdaemon Thank you for being patient. The examples helped. When this moves forward to action, it will take some careful rewriting of the introduction to bring into focus the original categories for return as not specify a value versus specify a value (even if `undefined` or `void` whatever). So that the option turns the categories to return `undefined` versus return anything else. Just to make sure I am finally on the page, return a variable which has the value `undefined` counts as anything else?

**Comment by scriptdaemon** [comment]
@pedrottimark When a user returns `undefined` or `void`, it's fairly clear as to the intent, and it is also consistent. Whereas a variable that just happens to be `undefined` may not be intentional in every case (and would thus be inconsistent).

For example, say a user has a global variable that isn't initialized until some action that happens later in the code:

```js
var example;
var example2;

function getExample(arg){
    if (!arg) {
        return 'Bad input';
    }
    if (arg === 'example') {
        return example; // Could be undefined, could be something else
    }
    return example2; // Could be undefined, could be something else
}
```

**Comment by nzakas** [comment]
@eslint/eslint-team we still need a champion for this (we have three 👍s)

**Comment by pedrottimark** [comment]
I will be champion to work through the rest of the process with @scriptdaemon

**Comment by ilyavolodin** [comment]
Ok, looks like we have everything we need. Marking as accepted.
